registry/finders.go

```diff
@@ -113,28 +122,125 @@ func FindAppAttachment(c *Space, appSlug, filename string, channel Channel) (*At
 }
 
 func FindVersionAttachment(c *Space, appSlug, version, filename string) (*Attachment, error) {
+	var headers swift.Headers
+	var md5Sum, contentType string
+	var fileContent []byte
+
 	// Return from swift
 	conf, err := config.GetConfig()
 	if err != nil {
 		return nil, err
 	}
 	sc := conf.SwiftConnection
-	var buf bytes.Buffer
+
+	var contentBuffer = new(bytes.Buffer)
 	fp := filepath.Join(appSlug, version, filename)
-	prefix := GetPrefixOrDefault(c)
-	headers, err := sc.ObjectGet(prefix, fp, &buf, false, nil)
-	if err != nil {
-		return nil, err
+
+	// First, we try to get the attachment from the global asset database.
+	ver, err := FindVersion(c, appSlug, version)
+
+	md5Sum, ok := ver.AttachmentReferences[filename]
+	if ok {
+		contentBuffer, headers, err = asset.AssetStore.FS.GetAsset(md5Sum)
+		fileContent, err = ioutil.ReadAll(contentBuffer)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// If we cannot find it, we try from the local database as a fallback
+		prefix := GetPrefixOrDefault(c)
+		headers, err = sc.ObjectGet(prefix, fp, contentBuffer, false, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		fileContent = contentBuffer.Bytes()
+
+		// We are going to move the asset to the global database and remove it
+		// from the local database for the next time
+		go func() {
+			err = MoveAssetToGlobalDatabase(c, ver, bytes.NewReader(fileContent), filename, contentType)
```

Comment:
Why do you create a `bytes.NewReader` instead of just passing `fileContent` to `MoveAssetToGlobalDatabase`?

Reply:
And `ver` is always nil here, no?